libbpf-tools/syscount.c

```diff
@@ -0,0 +1,482 @@
+{
+	printf("%-22s %8s\n", agg_colname(), "COUNT");
+}
+
+static void print_latency(struct data_ext_t *vals, size_t count)
+{
+	double div = env.milliseconds ? 1000000.0 : 1000.0;
+	char buf[2 * TASK_COMM_LEN];
+
+	print_latency_header();
+	for (int i = 0; i < count && i < env.top; i++)
+		printf("%-22s %8llu %16.3lf\n",
+		       agg_col(&vals[i], buf, sizeof(buf)),
+		       vals[i].count, vals[i].total_ns / div);
+	printf("\n");
+}
+
+static void print_count(struct data_ext_t *vals, size_t count)
+{
+	char buf[2 * TASK_COMM_LEN];
+
+	print_count_header();
+	for (int i = 0; i < count && i < env.top; i++)
+		printf("%-22s %8llu\n",
+		       agg_col(&vals[i], buf, sizeof(buf)), vals[i].count);
+	printf("\n");
+}
+
+static void print_timestamp()
+{
+	time_t now = time(NULL);
+	struct tm tm;
+
+	if (localtime_r(&now, &tm))
+		printf("[%02d:%02d:%02d]\n", tm.tm_hour, tm.tm_min, tm.tm_sec);
+	else
+		warn("localtime_r: %s", strerror(errno));
+}
+
+static bool batch_map_ops = true; /* hope for the best */
+
+static bool read_vals_batch(int fd, struct data_ext_t *vals, __u32 *count)
+{
+	struct data_t orig_vals[*count];
+	void *in = NULL, *out;
+	__u32 n, n_read = 0;
+	__u32 keys[*count];
+	int err = 0;
+
+	while (n_read < *count && !err) {
+		n = *count - n_read;
+		err = bpf_map_lookup_and_delete_batch(fd, &in, &out,
+				keys + n_read, orig_vals + n_read, &n, NULL);
+		if (err && errno != ENOENT) {
+			/* we want to propagate EINVAL upper, so that
+			 * the batch_map_ops flag is set to false */
+			if (errno != EINVAL)
+				warn("bpf_map_lookup_and_delete_batch: %s\n",
+				     strerror(-err));
+			return false;
+		}
+		n_read += n;
+		in = out;
+	}
+
+	for (__u32 i = 0; i < n_read; i++) {
+		vals[i].count = orig_vals[i].count;
+		vals[i].total_ns = orig_vals[i].total_ns;
+		vals[i].key = keys[i];
+		strncpy(vals[i].comm, orig_vals[i].comm, TASK_COMM_LEN);
+	}
+
+	*count = n_read;
+	return true;
+}
+
+static bool read_vals(int fd, struct data_ext_t *vals, __u32 *count)
+{
+	__u32 keys[MAX_ENTRIES];
+	struct data_t val;
+	__u32 key = -1;
+	__u32 next_key;
+	int i = 0;
+	int err;
+
+	if (batch_map_ops) {
+		bool ok = read_vals_batch(fd, vals, count);
+		if (!ok && errno == EINVAL) {
+			/* fall back to a racy variant */
+			batch_map_ops = false;
+		} else {
+			return ok;
+		}
+	}
+
+	if (!vals || !count || !*count)
+		return true;
+
+	for (key = -1; i < *count; ) {
+		err = bpf_map_get_next_key(fd, &key, &next_key);
+		if (err && errno != ENOENT) {
+			warn("failed to get next key: %s\n", strerror(errno));
+			return false;
+		} else if (err) {
+			break;
+		}
+		key = keys[i++] = next_key;
+	}
+
+	for (int j = 0; j < i; j++) {
+		err = bpf_map_lookup_elem(fd, &keys[j], &val);
+		if (err && errno != ENOENT) {
+			warn("failed to lookup element: %s\n", strerror(errno));
+			return false;
+		}
+		vals[j].count = val.count;
+		vals[j].total_ns = val.total_ns;
+		vals[j].key = keys[j];
+		strncpy(vals[j].comm, val.comm, TASK_COMM_LEN);
```

Comment:
newer GCCs will complain about potentially non-zero-terminated strings. I'd just do straight memcpy of TASK_COMM_LEN bytes (which we know are zero-terminated).

Reply:
ok, done